### ISSUE TITLE
Normalization for convolution layer weights

### ISSUE THREAD

**Issue by tarikarici** [BVLC/caffe#3242]
I have noticed that gradient w.r.t convolution layer parameters are summed over all pixel multiplications of previous layer feature image and next layer gradient image. However, feature image sizes (hidden layer images) change from layer to layer. 

So the gradient at each step is a sum of all gradients in the gradient image, without normalization. So, effectively, convolution layer weight-gradients used in stochastic-gradient-solver is using varying learning rates. Therefore, I think the backward methods require a normalization depending on the number of pixels in the gradient.

**Comment by seanbell** [comment]
The expression that is computed in caffe is the correct gradient of the parameters with respect to the loss function.  What you're describing is using something other than the gradient to update the parameters.

For example, the RMSProp solver doesn't use the gradient -- it has a moving average that behaves a little bit like using only the sign of the gradient.  You can also interpret it as having a separate "learning rate" per parameter.

What might be better is a count of the number of times a parameter is touched.  Then the solver could use this information and down-weight gradients by this value.   I think part of the reason why RNNs can be unstable to train is that some parameters accumulate way more terms than others, and this is why RMSprop works well for those models.

**Comment by tarikarici** [comment]
I have a good understanding of how the gradient is back-propagated for inner product layers but I haven't read anything on convolution layers. Now I am planning to read on it :). 

But I still think convolution weights have a learning rate that is the number-of-pixels-in-the-next-layer times more than it should be. My argument: normally each next-layer-pixel should be a node (unit) but their weights are shared across all the pixels so it boils down to convolution. Any weight gradient is: output-of-prev-layer times times gradient-of-next-layer. Since these weight gradients are different, to compute the update of a weight in the convolution we should average all gradients. Not accumulate them. Hence we are projecting the gradients to parameter sharing constraint. This accumulation is done via matrix multiplication.

So, according to me, effectively convolution layer weights have a larger learning rate depending on the pixels in the next layers output. This might not necessarily be bad since convolution layers are generally earlier than inner product layers and so effectively inner product layer learning rates are smaller. This might help with our stochastic-wandering-around optimization since changing weights in the later layers might confuse our gradient directions in the earlier layers.

I like your idea that the number of times a parameter is touched should effect the learning rate.

**Comment by seanbell** [comment]
I think you're confusing the concepts of gradient (derivative of loss with respect to parameter) with the optimal update for each parameter.  The expression in caffe is the correct gradient.  If doing SGD, you should use the gradient.  Each individual weight has _N_ times more terms because it has _N_ times the effect on the loss output.

**Comment by tarikarici** [comment]
Thanks for the reply. And I really hate taking your time but :).... And I know CNN BP is well established and as I now see, not just Caffe but everybody does it the same way by accumulation.

I read the following tutorial: http://cogprints.org/5869/1/cnn_tutorial.pdf

In this tutorial it says in quote: "Finally, the gradients for the kernel weights are computed using backpropagation, except in this case the same weights are shared across many connections. We’ll therefore sum the gradients for a given weight over all the connections that mention this weight, ..."

First, each product of previous layer output and next layer gradient is the gradient of the weight. And to share the weights across all convolutions in an image, gradients are summed. I think this is the problem. This is a projection on to the weight-sharing-constraint and the way to do it is by averaging. Not summation. This effectively sums the corresponding learning rates for convolution layer in network by N.

**Comment by seanbell** [comment]
Summing vs averaging is not a design choice.  It's not a question of "who does it what way?"  It is what you get if you derive the gradient using calculus.

Consider `L(f(x), g(x), h(x))`.  The gradient for `x` is `dL/dx`, which is `dL/df * df/dx + dL/dg * dg/dx + dL/dh * dh/dx`.  Now consider the case where `f(x) = g(x) = h(x) = x`, i.e. the variable `x` is used multiple times inside `L`.  Then `df/dx = dg/dx = dh/dx = 1` and we get the expression `dL/dx = dL/df + dL/dg + dL/dg`.  Each variable `f, g, h` corresponds to one place in which `x` was used inside `L`.  And the result is a sum, derived from first principles.  You should try deriving the expression yourself for the case of a convolution.

**Comment by tarikarici** [comment]
Sean, thanks for the reply. Last night, after replying to you, I wrote an email to Prof. Geoffrey Hinton describing what I have claimed above. And this is his reply: 

---
people are very well aware of this.

 Geoff
---

A short one, but still he replied :)

So what I said was true and effectively learning rate for convolution layer parameters is effectively N times larger, where N is the number of pixels in the next layer.

To respond to your argument, I have no problem with the chain rule. In fact the change rule says a single weight gradient is previous_layer_output * next_layer_gradient, just like it is with inner product. But there are not N times the same parameter. There are N different parameters and they are constrained to be equal that is what you are missing. I will try to explain more precisely below:

To get a convolution layer, think of each output pixel as a single unit in the next layer, but a "locally" connected one. So we have N of these units. And we can backpropagate and find the gradient for each weight in the local support for that next layer pixel. Let's say it is a 3x3 filter. So we can backpropagate and find 3x3=9 weight gradients for each N pixel (unit). But we want all of these N weights to be equal as we update via SGD. So we are going to start with the same initials and make their gradients equal. To do this, we must average them (as it is a projection operation on to an equality constraint). But in almost all implementations that I can see the N gradients are summed not averaged. And this effectively multiplies the corresponding weights by N.

But as I said earlier this may have helped in CNNs simply because as we go back in the network, gradients become smaller so we are scaling them up or changing later layers like inner-product layers slowly stabilizes the convergence.

But I think this is still problematic. Because then the learning rate depends on the filter-mask size and padding which determines the number of pixels in the output. Also, pooling becomes a little confusing as the way it functions becomes complicated. Pooling not only functions its usual role but also effectively determines the learning rate. And I know learning rate is really important in training CNNs. I think this issue requires a good analysis and research. Maybe there is and I am not yet aware of. I think there is still a lot of room for improvement in SGD part.

**Comment by seanbell** [comment]
You say:
>  To do this, we must average them (as it is a projection operation on to an equality constraint). But in almost all implementations that I can see the N gradients are summed not averaged. And this effectively multiplies the corresponding weights by N.

Can you derive how projecting onto an equality constraint results in the average?  I understand that intuitively it feels like the right choice, but I'm not aware of any theoretical justification for it.

**Comment by tarikarici** [comment]
Let me write the optimization problem

min Loss(w1,w2) over w1, w2 such that w1 = w2

If we need to find the gradient and project on to the constraint set successively (just like Bregman's algo).

By taking the derivative of the loss I can find $\delta_1$ and $\delta_2$ as derivative w.r.t to $w_1$ and $w_2$.  Let's call this $d$. And the constraint vector, I want to project onto is $c=[1 1]$. Hence linear projection operation based on Euclidean distance is $(d^T*c/|c|) c/|c|$ which is $\hat{d}=[(\delta_1 + delta_2)/2,  (\delta_1 + delta_2)/2 ]$. 

So we keep iterating above computations.


Now

**Comment by seanbell** [comment]
I see what you mean.  You could also write it using a Lagrangian:
`Loss(w1, w2) + lambda * ||w1 - w2||^2`

Hinton's "Dark Knowledge" method and some papers from LeCun's group propose methods somewhat like this, but where the constraints are across different CNNs and not within a CNN like what you are proposing.

It might do better or worse than SGD which uses the gradient of `w` after collapsing `w = w1 = w2`.  Hard to say without trying it.  I can see it doing better if you leave all the variables separate and regularize them to be close to each other, because this introduces freedom in the optimization problem.  But it would also use up a huge amount of RAM and may not be feasible for the best-performing models.

**Comment by tarikarici** [comment]
I don't think it is a good idea that we use regularization instead of constraints and divert away from convolution. Then we will be increasing the degrees of freedom (dof) by the number of pixels in the image, which is huge. I am happy with convolution layers at earlier stages and then switching to inner product layers when you want more connections between features to learn intricate relationships. And if I want more dof for my model, I would simply increase the number of features in the convolution layer, rather than letting filters change spatially. 

But my point about this issue is important I think. If we do not average the gradients, then there is a big difference between the learning rate of convolution layer and inner product layers. And this difference depends on filter mask size, padding, number of pooling in between that is the network structure. Learning rate of layers changes by a big amount depending on the input image size. This is the reason why we have to choose a really small learning rate for bigger images. If updating convolution layers with a bigger rate is useful this should be done explicitly not behind our back.

**Comment by jeffdonahue** [comment]
I'm closing as this is not a Caffe issue.  I'll echo @seanbell here -- at the end of the day, `Backward` has a simple mathematical definition: it computes the gradient of the error (i.e., the function computed by `Forward`) with respect to the parameters, and that's what Caffe does (modulo any actual bugs...). There might be good intuition here and it's possible this is in fact a better way of training, and I don't want to discourage @tarikarici or anyone else from giving it a shot*, but the behavior of Caffe's `Backward` is mathematically correct as is.

*If you wanted to try this on an existing net, you could do so without modifying any code just by setting `param { lr_mult: x }`, where `x` is the appropriate scaling factor (maybe `x := kernel_size^(-2)` if I'm reading this correctly), in each `Convolution` layer. Or of course you could hack the `Convolution::Backward` code itself if you prefer.

**Comment by tarikarici** [comment]
Actually my whole point was Backward is not computing the correct gradient but N times the correct gradient, and this N changes with the network.

Yes lr_mult can be used as you said to correct the problem via adjusting the learning rates.

**Comment by seanbell** [comment]
I certainly encourage you to try using that as your update rule, but what is computed in caffe is the correct gradient (i.e. the derivative of the loss with respect to each free parameter).  You can test this with finite difference (caffe does this in its unit tests) and you will see that the current definition is correct.

That being said, you don't have to use the gradient to update the parameters.  You could scale the gradient based on the criteria that you're proposing, and it might do better.  If you think about optimization as point moving towards an optimum, in the worst case the gradient point a 90 degree angle to the correct solution, and your re-scaling might reduce this angle.

**Comment by tarikarici** [comment]
I thought we agreed with you on the gradient :). Sum can not be the gradient. Probably the test also sums up all gradients and compares the sum. My guess...

**Comment by seanbell** [comment]
If you're curious, this is the utility code that checks the gradient: https://github.com/BVLC/caffe/blob/master/include/caffe/test/test_gradient_check_util.hpp

Here is one place where the gradient checker is instantiated for convolution layers:
https://github.com/BVLC/caffe/blob/master/src/caffe/test/test_convolution_layer.cpp#L624

The gradient checker compares individual scalar values; there is nothing to sum up since you're comparing just one output (some value in a `top`) to one parameter (some value in a parameter `blob`).
